clash-cores/src/Clash/Cores/UART.hs

```diff
@@ -0,0 +1,207 @@
+{-|
+  Copyright   :  (C) 2021, LUMI GUIDE FIETSDETECTIE B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  Rowan Goemans <goemansrowan@gmail.com>
+
+  UART transmitter and receiver module
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE NumericUnderscores #-}
+
+module Clash.Cores.UART(HzToPeriod, uartTx, uartRx, uart) where
+
+import Clash.Prelude
+
+-- | Convert frequency in hertz to period in picoseconds
+type HzToPeriod (rate :: Nat) = Div 1_000_000_000_000 rate
```

Comment:
I agree there are multiple cases where I have used this. A PWM module, An SDRAM module and more...

Reply:
The first I saw of it was in a [message from Érdi Gergő](https://groups.google.com/g/clash-language/c/vmPHLqdy4rQ/m/feoTrfr7CAAJ?utm_medium=email&utm_source=footer). We ended up [making hzToPeriod round exactly the same](https://groups.google.com/g/clash-language/c/DYKLt_v_Xjg/m/PYgXRl31AQAJ), for least‍-‍surprising behaviour. I think adding it to Prelude makes a lot of sense.

Reply:
I agree but I think that should be a seperate PR.

Reply:
Somewhat unrelated, but I also have

```haskell
data TimeUnit = Ps | Ns | Us | Ms | Sec | Min | Hr

type Ps  = 'Ps
type Ns  = 'Ns
type Us  = 'Us
type Ms  = 'Ms
type Sec = 'Sec
type Min = 'Min
type Hr  = 'Hr

-- Turn it into picoseconds
type FromUnit :: TimeUnit -> Nat -> Nat
type family FromUnit u n where
  FromUnit 'Ps  n = n
  FromUnit 'Ns  n = n * 1000
  FromUnit 'Us  n = n * 1000000
  FromUnit 'Ms  n = n * 1000000000
  FromUnit 'Sec n = n * 1000000000000
  FromUnit 'Min n = n * 60000000000000
  FromUnit 'Hr  n = n * 600000000000000
```

So I can write things like

```haskell
FromUnit Sec 10
```

instead of actually bothering to figure out how long some amount of time is. Maybe this is also useful?

Reply:
I have [a simpler version of this](https://github.com/gergoerdi/retroclash-lib/blob/668e723bda32ec689600dd17574b1ac5cc008fa5/src/RetroClash/Clock.hs#L19-L25), is there any benefit to a separate `FromUnit` type family?

```
type Seconds      (s  :: Nat) = Milliseconds (1_000 * s)
type Milliseconds (ms :: Nat) = Microseconds (1_000 * ms)
type Microseconds (us :: Nat) = Nanoseconds  (1_000 * us)
type Nanoseconds  (ns :: Nat) = Picoseconds  (1_000 * ns)
type Picoseconds  (ps :: Nat) = ps
```

Reply:
> is there any benefit to a separate FromUnit type family

No, this is just the first thing I happened to write. Then I threw it into a _bag-of-miscellany_ module and didn't look at it again

Reply:
I'm resolving this change requesst for now, if and when we add `HzToPeriod` to Prelude, we can change the UART code to use it then.